src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -1985,53 +2970,194 @@ public UncaughtExceptionHandler getUncaughtExceptionHandler() {
+            NOT_SUPPORTED_CLASSLOADER = loader;
+        }
+    }
+
+    /**
+     * Returns the special ThreadGroup for virtual threads.
+     */
+    static ThreadGroup virtualThreadGroup() {
+        return Constants.VTHREAD_GROUP;
+    }
+
     // The following three initially uninitialized fields are exclusively
     // managed by class java.util.concurrent.ThreadLocalRandom. These
     // fields are used to build the high-performance PRNGs in the
-    // concurrent code, and we can not risk accidental false sharing.
-    // Hence, the fields are isolated with @Contended.
+    // concurrent code.
 
     /** The current seed for a ThreadLocalRandom */
-    @jdk.internal.vm.annotation.Contended("tlr")
     long threadLocalRandomSeed;
 
     /** Probe hash value; nonzero if threadLocalRandomSeed initialized */
-    @jdk.internal.vm.annotation.Contended("tlr")
     int threadLocalRandomProbe;
 
     /** Secondary seed isolated from public ThreadLocalRandom sequence */
-    @jdk.internal.vm.annotation.Contended("tlr")
     int threadLocalRandomSecondarySeed;
 
+    /** The thread container that this thread is in */
+    private volatile ThreadContainer container;  // @Stable candidate?
+    ThreadContainer threadContainer() {
+        return container;
+    }
+    void setThreadContainer(ThreadContainer container) {
+        // assert this.container == null;
+        this.container = container;
+    }
+
+    /** The top of this stack of stackable scopes owned by this thread */
+    private volatile StackableScope headStackableScopes;
+    StackableScope headStackableScopes() {
+        return headStackableScopes;
+    }
+    static void setHeadStackableScope(StackableScope scope) {
+        currentThread().headStackableScopes = scope;
+    }
+
     /* Some private helper methods */
-    private native void setPriority0(int newPriority);
-    private native void stop0(Object o);
-    private native void suspend0();
-    private native void resume0();
-    private native void interrupt0();
-    private static native void clearInterruptEvent();
-    private native void setNativeName(String name);
+    // The address of the next thread identifier, see ThreadIdentifiers.
+    private static native long getNextThreadIdOffset();
+
+    private void setPriority0(int newPriority) {
+        synchronized (interruptLock) {
+            if (started && (NO_REF != eetop)) {
+                setPriorityNoVMAccessImpl(eetop, newPriority);
+            }
+        }
+    }
+
+    private void stop0(Object o) {
+        synchronized (interruptLock) {
+            if (o != null) {
+                if (started) {
+                    /* [PR CMVC 179978] Java7:JCK:java_lang.Thread fails in all platform */
+                    /*
+                        * If the thread has not yet been simply store the fact that stop has been called,
+                        * the JVM uses this to determine if stop has been called before start.
+                        */
```

Comment:
Please correct the indentation of these lines.

Reply:
Updated.